### ISSUE TITLE
GroupBy date parts

### ISSUE THREAD

**Issue by kevinob11** [balderdashy/sails-postgresql#164]
I'd like to GroupBy date parts, for example:
```
Invoice.find().sum('total').groupBy('date').exec();
```
This "works" but groups by each individual date and time, so you don't actually get any groups. If I want to group by month or day I need to be able to do something like:
```
Invoice.find().sum('total').groupBy("date_trunc( 'month', invoice.date)").exec();
```
Which, as expected, doesn't work. I can of course write a custom query, but then I lose all of the nice automatic where handling etc. that I get with the query builder.

Thanks!

**Comment by devinivy** [comment]
Currently the sequel-builder for Waterline adapters doesn't support raw sequel in groupBy.  I don't _think_ it would require any more than a change to waterline-sequel, around [here](https://github.com/balderdashy/waterline-sequel/blob/d1f1b51c5f6ffdb118275250994607dba140d506/sequel/lib/criteriaProcessor.js#L861).  What do you think about allowing for this sort of thing, @dmarcelino ?  On one hand, it could become part of the SQL interface.  On the other hand, maybe it should be discouraged since it's impossible to support across adapters.  And lastly, this all makes me wonder if the fields sent to groupBy get transformed from attribute names to column names, or if they're currently expected to be column names.  The former _should_ be true, but I'm not sure if that's tested.

**Comment by dmarcelino** [comment]
> What do you think about allowing for this sort of thing, @balderdashy/dmarcelino?

I'd be OK with that. The same way adapters can extend waterline standard behaviour with custom methods I don't see much harm in extending the supported query syntax, assuming it's documented on the README.

**Comment by kevinob11** [comment]
Looking at waterline-sequel I believe it just expects they are column names. Rather than changing default groupBy behavior I forked and quickly added a groupByDate as a simple proof of concept.

https://github.com/praece/waterline-sequel/blob/master/sequel/lib/criteriaProcessor.js#L881

This could of course be generalized to just be groupByQuery or something that signifies that it allows a plain sql string. It also needs to be deal with the fact that groupByDate and groupBy can't both be used at once right now because they both add the groupBy clause.

Is this what you @dmarcelino mean by custom methods? Something that doesn't pollute the default groupBy method but adds the feature for those adapters that can support it?

**Comment by dmarcelino** [comment]
I was actually hoping that it would be possible to modify `groupBy` so it takes something that is not a column such as raw sql...

**Comment by kevinob11** [comment]
Ah I see. Something like if it is a string see if it matches a known column name for this table if not assume it is raw sql?

**Comment by dmarcelino** [comment]
Yeap!

**Comment by kevinob11** [comment]
Cool, I think I could give that a shot today or tomorrow, I'll report back.

**Comment by kevinob11** [comment]
Ok at the link below you'll find an if that checks the current schema for a column matching the value submitted in the groupBy. If it matches the value it assumes you want to group on a simple column, if not it assumes you are sending in a complicated expression and will just add it to the query. I took care of adding it to the select statement as well.

I also found an issue in that when you add it to the select statement the automatic naming of the output is pretty cruddy, with values like `to_char`. The problem is that you can't pass in an `as` because it breaks the actual group by statement. As a simple fix I just name each `group#` where # is the index of the group by array if there are multiple items to group on. I'm open to suggestions to make this better.

https://github.com/praece/waterline-sequel/blob/master/sequel/lib/criteriaProcessor.js#L871

Thoughts? If this all sounds sane I'll try my hand at some tests!

**Comment by dmarcelino** [comment]
Looks good to me. I think this is reasonable solution.

Tests would be nice indeed!

**Comment by kevinob11** [comment]
My week got away from me, I'll get those written this week.

Somewhat related what do you think about aggregates (min, max, sum, groupBy) allowing you to pass in a second parameter to give the aggregate a name? It would solve my group by problem of crappy naming and would solve the issue that we appear to have that you can't get both the min and max of a column because they end up with the same name. If this also sounds reasonable I'll open a second ticket and start that as well.

**Comment by kevinob11** [comment]
I'm realizing accepting that second parameter is not a waterline-sequel thing, and would be a big change for waterline, which likely needs more thought and much more input from elsewhere before being considered. I'll stick to the tests and try to sort out the other issue some other way.

**Comment by kevinob11** [comment]
Ok I wrote some tests and finally opened the pull request. Let me know if any changes are required!

**Comment by sailsbot** [comment]
Thanks for posting, @kevinob11.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by kevinob11** [comment]
Should we keep this open until merged? Or managed through the pull request. Has something changed since I made the pull request?

**Comment by devinivy** [comment]
This one is totally open!

**Comment by kevinob11** [comment]
Any comments or changes needed on the pull request? I'm happy to make changes as necessary!

**Comment by sailsbot** [comment]
Thanks for posting, @kevinob11.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by kevinob11** [comment]
@devinivy @dmarcelino this was auto-closed again. I'd love to get it merged and really close it up.

**Comment by sailsbot** [comment]
Thanks for posting, @kevinob11.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
Links to: https://github.com/balderdashy/waterline-sequel/pull/59

**Comment by kevinob11** [comment]
I'd love to get this merged so I don't have to have my own fork in my app, anything else I can do from my end?

**Comment by sailsbot** [comment]
Thanks for posting, @kevinob11.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by kevinob11** [comment]
Can we merge this?!

**Comment by particlebanana** [comment]
published waterline-sequel `0.5.2` which includes this fix
